compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/PromisedValue.kt

```diff
@@ -70,6 +77,30 @@ abstract class PromisedValue(val codegen: ExpressionCodegen, val type: Type, val
 
     val typeMapper: IrTypeMapper
         get() = codegen.typeMapper
+
+    private fun IrType.isJava() =
+        (this.classifierOrFail.owner as IrDeclaration).origin == IrDeclarationOrigin.IR_EXTERNAL_JAVA_DECLARATION_STUB
+
+    private fun IrType.isPackagePrivate(): Boolean {
+        return this.getClass()?.visibility?.delegate == JavaVisibilities.PackageVisibility
+    }
+
+    private fun IrType.inDifferentPackage(): Boolean =
+        this.getClass()!!.getPackageFragment()?.fqName != codegen.irFunction.getPackageFragment()?.fqName
+
+    private fun computeCoerceTarget(
+        isUpcastToJava: Boolean,
+        erasedTargetType: IrType,
+        default: Type
+    ): Type {
+        if (!isUpcastToJava && erasedTargetType.isJava() && erasedTargetType.isPackagePrivate() && erasedTargetType.inDifferentPackage()) {
+            val computed = irType.classifierOrNull?.superTypes()?.firstOrNull {
```

Comment:
Note that `superTypes` returns only immediate supertypes, and we probably need all supertypes. Please add a test where all immediate supertypes of a package-private class in other package are also package-private in that package, to check that we consider all supertypes here

Reply:
Thanks for reminding me of this, I have made a little modification, instead of looking for a `public super type`  of `erasedSourceType` as earlier, a `visible super type` should be good enough. if `T <: S` where `S` is `package-private` java type, we can always find a visible super type `K` in the immediate supertypes of `T`. I have managed to add a test to cover this, hope I understand what you mean correctly.